src/crewai/agents/executor.py

```diff
@@ -185,6 +197,52 @@ def _iter_next_step(
             yield AgentStep(action=output, observation=observation)
             return
 
+        except BadRequestError as e:
+            print("Bad Request Error", e)
+            if "context_length_exceeded" in str(e):
+                self._logger.log(
+                    "debug",
+                    "Context length exceeded. Asking user if they want to use summarize prompt to fit, this will reduce context length.",
+                    color="yellow",
+                )
+                user_choice = prompt(
+                    "Context length exceeded. Do you want to summarize the text to fit models context window? (Y/N): ",
+                    validator=YesNoValidator(),
+                ).lower()
+                if user_choice in ["y", "yes"]:
+                    self._logger.log(
+                        "debug",
+                        "Context length exceeded. Using summarize prompt to fit, this will reduce context length.",
+                        color="bold_blue",
+                    )
+                    intermediate_steps = self._handle_context_length(intermediate_steps)
+
+                    output = self.agent.plan(
+                        intermediate_steps,
+                        callbacks=run_manager.get_child() if run_manager else None,
+                        **inputs,
+                    )
+
+                    if isinstance(output, AgentFinish):
+                        yield output
+                    else:
+                        yield AgentStep(action=output, observation=None)
+                    return
+                else:
+                    raise SystemExit(
+                        "Context length exceeded and user opted not to summarize. Consider using smaller text or RAG tools from crewai_tools."
+                    )
+
+            else:
+                raise e
+
+        except Exception as e:
+            yield AgentStep(
+                action=AgentAction("_Exception", str(e), str(e)),
+                observation=str(e),
+            )
+            return
+
```

Comment:
_Suggestion_: All this code could be its method to handle these cases. As I suggested above, if you have a Custom exception you could do something like this:
_hint_: If we have branched out to a specific case and it has too many lines, might be better to have it in a different function to encapsulate and isolate that behavior to be more testable and easier to read

```
except ContextLengthException as exc:
     self._handle_conext_length_exc()
```